src/lib.rs

```diff
@@ -44,6 +46,19 @@ pub use middleware::GrantsMiddleware;
 ///     HttpResponse::Ok().body("some secured info")
 /// }
 ///
+/// // Custom access denied message.
+/// #[has_roles("ADMIN", error = "access_denied()")]
```

Comment:
Let's slightly change the signature for the parameter.
What do you think of passing only `fn` name without `()`?

Reply:
Yes, I was debating the same thing! Something similar to `serde` - `#[serde(deserialize_with = "path")]` and an accompanying function

```
fn path<'de, D>(deserializer: D) -> Result<ActualData, D::Error>
where
    D: de::Deserializer<'de>,
{
    ...
}
```
But 

1, I wasn't sure what param(s) this function should receive 
2. or if it should require any param(s) at all because the purpose of that function should be constrained to setting status code and response body
3. maybe then `error = "fn()"` (function with empty parenthesis) indicates the user that this function shouldn't require any param(s) 
4. Or maybe `error = "expr"` should just be a const expression for response body since the status code should always be `FORBIDDEN` (403)

`fn` without `()` does look cleaner then the 3rd option so should I finalize that or should I consider the 4th option?

Reply:
Well, that's really good points.
Btw, just like in serde, this approach is used by `actix-web` (e.g: `guard="some_fn"`) and other crates.

I thought about possible arguments for this `fn`, I'm not sure if it useful for users of the crate. There might be a feature request later.

Actually, `ident` looks more restrictive than `expr`. Look at the following example with `expr` approach:
```rust
fn access_denied_str(body: &'static str) -> HttpResponse {
    HttpResponse::with_body(
        StatusCode::FORBIDDEN,
        BoxBody::new(body),
    )
}

#[get("/access")]
#[has_roles("ADMIN", error = "access_denied_str(\"Example\")")]
async fn access_response() -> &'static str {
    "Hi!"
}
```
And it has advantages and disadvantages.
Pros: 
1) One function can be used for all endpoints with different custom arguments (not sure if it's convenient).
2) Users can use this in a lot of ways and passing thier own arguments. Moreover it might be constants or smtg else.

Cons: 
1)it's harder to maintain backward capability in new versions. 
2) Moreover it restricts us to inject params to such `fn`s automatically (e.g: `AuthDetails` or `ServiceRequest`).

It seems to me that it's better to introduce new features with stricter approach and expand as users request. So `ident` seems to be a better start. What do you say?

Reply:
> It seems to me that it's better to introduce new features with stricter approach and expand as users request.

Agreed.
I don't think users will want this library to be able to send elaborate response body, I believe that would fall more in front-end, web-server domain. I think we could potentially restrict current implementation to a simple `&str`, like this:

``` rust
=  46 const NON_ADMIN_RESPONSE: &str = "This resource allowed only for ADMIN";
   47
   48 #[get("/access")]
=  49 #[has_roles("ADMIN", on_forbidden = "NON_ADMIN_RESPONSE")]
   50 async fn access_response() -> &'static str {
   51 │   "Hi!"
   52 }
```
setting a `const &str` (instead of `on_forbidden="a very long message ..."`) could also be easier to read if the message was multi-lined.

And in

`expand.rs`
``` rust
 =  83 │   │   let resp = if let Some(expr) = &self.args.on_forbidden {
    84 │   │   │   quote!(#expr)
    85 │   │   } else {
 =  86 │   │   │   quote!("")
    87 │   │   };
    88
    89 │   │   let stream = quote! {
    90 │   │   │   #(#fn_attrs)*
    91 │   │   │   #func_vis #fn_async fn #fn_name #fn_generics(
    92 │   │   │   │   _auth_details_: actix_web_grants::permissions::AuthDetails<#type_>,
    93 │   │   │   │   #fn_args
    94 │   │   │   ) -> actix_web::Either<#fn_output, actix_web::HttpResponse> {
    95 │   │   │   │   use actix_web_grants::permissions::{PermissionsCheck, RolesCheck};
 +  96 │   │   │   │   use actix_web::{http::StatusCode, body::BoxBody};
    97 │   │   │   │   #condition {
    98 │   │   │   │   │   let f = || async move #func_block;
    99 │   │   │   │   │   actix_web::Either::Left(f().await)
   100 │   │   │   │   } else {
 = 101 │   │   │   │   │   actix_web::Either::Right(HttpResponse::with_body(StatusCode::FORBIDDEN, BoxBody::new(#resp)))
   102 │   │   │   │   }
   103 │   │   │   }
   104 │   │   };
```
followed by

``` rust
  110 struct Args { 1 implementation
  111 │   permissions: Vec<syn::LitStr>,
  112 │   secure: Option<syn::Expr>,
  113 │   type_: Option<syn::Expr>,
= 114 │   on_forbidden: Option<syn::Ident>,
  115 }		
```

I am not totally sure on how compiler optimization works with macros and the performance but perhaps this could also be easier for the compiler to optimize away, instead of an `fn` call, just inline.

Reply:
While I'm guessing the main use case would be essentially body customization, I think a lot of people use `JSON` (or some other type) and they can use different status codes, so we must provide the ability to specify `fn`.

But I ike your suggestion 👍  This can be a separate parameter (`on_forbidden`, `error_body`, `forbidden_body`). You could do it even in separate PR 🤔

Reply:
Makes perfect sense. My latest commit declares `error` required `fn` as `fn() -> HttpResponse` and without parenthesis.

Leaving this
``` rust
let resp = if let Some(expr) = &self.args.error_fn {
            quote!(#expr())
        } else {
            quote!(actix_web::HttpResponse::Forbidden().finish())
        };
```
to be potentially extensible.

I'd be happy to make any additional changes if required, so just let me know.